### PR TITLE
Feature/enable replication graph

### PR DESCRIPTION
**Pull request by m-samiec**
GDK PR: https://github.com/spatialos/UnrealGDK/pull/1865

### CODE REVIEW

File: Game/Source/GDKTestGyms/GDKTestGyms.Build.cs [hunk 1]
```diff
@@ -14,13 +14,15 @@ public GDKTestGyms(ReadOnlyTargetRules Target) : base(Target)
...
+        "EngineSettings",
 				"GameplayAbilities",
 				"GameplayTasks",
 				"InputCore",
 				"Sockets",
 				"OnlineSubsystemUtils",
 				"PhysXVehicles",
-				"SpatialGDK"
+				"SpatialGDK",
+        "ReplicationGraph"
```
Comment (mironec) [inline Game/Source/GDKTestGyms/GDKTestGyms.Build.cs]: Weird spacing here and above

File: Game/Config/DefaultEngine.ini [hunk 2]
```diff
@@ -39,6 +39,9 @@ bAllowClientSideNavigation=True
 TargetedHardwareClass=Desktop
 DefaultGraphicsPerformance=Maximum
 
+[/Script/OnlineSubsystemUtils.IpNetDriver]
+;ReplicationDriverClassName=/Script/ReplicationGraph.BasicReplicationGraph
```
Comment (mironec) [inline Game/Config/DefaultEngine.ini]: why is this required?

File: Game/Source/GDKTestGyms/GDKTestGyms.Build.cs [hunk 3]
```diff
@@ -14,13 +14,15 @@ public GDKTestGyms(ReadOnlyTargetRules Target) : base(Target)
 				"Core",
 				"CoreUObject",
 				"Engine",
+        "EngineSettings",
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/GDKTestGyms.Build.cs]:
```suggestion
				"EngineSettings",
```

File: Game/Source/GDKTestGyms/GDKTestGyms.Build.cs (hunk 1 above)
Comment (m-samiec) [inline Game/Source/GDKTestGyms/GDKTestGyms.Build.cs]:
```suggestion
				"ReplicationGraph"
```

File: Game/Source/GDKTestGyms/Public/TestGymsReplicationGraph.h [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+private:
+	
+	TArray<FActorRepListRefView> ReplicationActorLists;
+	FActorRepListRefView ForceNetUpdateReplicationActorList;
+};
+
+/** This is a specialized node only used in Spatial to ensure that connection views and targets are replicated each frame. */
+UCLASS()
+class UTestGymsReplicationGraphNode_GlobalViewTarget : public UReplicationGraphNode
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Public/TestGymsReplicationGraph.h]: @improbable-danny @mironec This is a new node that has been added just for spatial

File: Game/Source/GDKTestGyms/Public/TestGymsReplicationGraph.h [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+	RelevantAllConnections,			// Routes to an AlwaysRelevantNode or AlwaysRelevantStreamingLevelNode node
+	
+	// ONLY SPATIALIZED Enums below here! See UTestGymsReplicationGraph::IsSpatialized
+
+	Spatialize_Static,				// Routes to GridNode: these actors don't move and don't need to be updated every frame.
+	Spatialize_Dynamic,				// Routes to GridNode: these actors mode frequently and are updated once per frame.
+	Spatialize_Dormancy,			// Routes to GridNode: While dormant we treat as static. When flushed/not dormant dynamic. Note this is for things that "move while not dormant".
+};
+
+/** TestGyms Replication Graph implementation. Based on UShooterReplicationGraph */
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Public/TestGymsReplicationGraph.h]: @mironec @improbable-danny Like it says, this is heavily based on the UShooterReplicationGraph

File: Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp [hunk 6]
```diff
@@ -0,0 +1,754 @@
...
+UTestGymsReplicationGraph::UTestGymsReplicationGraph()
+{
+	static ConstructorHelpers::FClassFinder<AActor> ReplicatedBP(TEXT("/Game/Actors/ReplicatedActor"));
+	if (ReplicatedBP.Class != nullptr)
+	{
+		ReplicatedBPClass = ReplicatedBP.Class;
+	}
+}
+
+void InitClassReplicationInfo(FClassReplicationInfo& Info, UClass* Class, bool bSpatialize, float ServerMaxTickRate)
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp]: @mironec @improbable-danny In this file, worth looking at the bits where functionality differs according to the `UsesSpatialNetworking` flag

File: Game/Config/DefaultEngine.ini (hunk 2 above)
Comment (m-samiec) [reply Game/Config/DefaultEngine.ini]: So this is used to set which replication class implementation you want to use. I've commented it out, as I don't want to enable it by default (not currently anyway).

File: Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp [hunk 7]
```diff
@@ -0,0 +1,754 @@
...
+
+	// ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	// Programatically build the rules.
+	// ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	
+	auto AddInfo = [&]( UClass* Class, EClassRepNodeMapping Mapping) { ClassRepNodePolicies.Set(Class, Mapping); };
+
+	AddInfo( APlayerState::StaticClass(),							EClassRepNodeMapping::NotRouted);				// Special cased via UTestGymsReplicationGraphNode_PlayerStateFrequencyLimiter
+	AddInfo( AReplicationGraphDebugActor::StaticClass(),			EClassRepNodeMapping::NotRouted);				// Not supported
+	AddInfo( AInfo::StaticClass(),									EClassRepNodeMapping::RelevantAllConnections);	// Non spatialized, relevant to all
```
Comment (mironec) [inline Game/Source/GDKTestGyms/Private/TestGymsReplicationGraph.cpp]: The naming collision between spatialized actors and Spatial as our things is bothering me a lot.

File: Game/Config/DefaultEngine.ini [hunk 8]
```diff
@@ -39,6 +39,9 @@ bAllowClientSideNavigation=True
 TargetedHardwareClass=Desktop
 DefaultGraphicsPerformance=Maximum
 
+[/Script/OnlineSubsystemUtils.IpNetDriver]
```
Comment (m-samiec) [inline Game/Config/DefaultEngine.ini]: Add comment here

**Comment by ernopp** [comment]
is it expected that there's no changelog entry here?

**Comment by m-samiec** [comment]
> is it expected that there's no changelog entry here?

Testgyms is an internal project - there's no changelog
